Optimize shallow_clone for Bytes::split_{off,to}

stepancheg opened the pull request:
If `shallow_clone` is called with `&mut self`, and `Bytes` contains
`Vec`, then expensive CAS can be avoided, because no other thread
have references to this `Bytes` object.

`split_off_and_drop` benchmark becomes 5% faster on my host (MacBook Pro, 2,2 GHz Intel Core i7).

This is related to #88.

carllerche commented:
Interesting. It looks like most of the work done in `shallow_clone` including the atomic load above. Maybe we could split this into two fns:

```rust
fn shallow_clone(&mut self) {
// ...
}

fn shallow_clone_sync(&self) {
....
}
```

Anything in common could maybe be broken out.

stepancheg commented:
It is not easy to split these function avoiding-copy-paste and keeping function readable.

Why would you want to split function in two? If you are worried about runtime overhead of branching, it could be replaced with function type parameter. [Here](https://github.com/carllerche/bytes/pull/92/commits/1faea2568abf82e13895a5977aa20f8eaf544353).

arthurprs commented:
This would be way nicer with an extra `exclusive: bool` parameter instead. The compiler will happily inline it (probably even without #[inline] annotations) and remove the branch.

carllerche commented:
Seems like the branch is out of date. If @arthurprs is correct, then the benchmarks should show it and the cleaner code is preferable 👍

stepancheg commented:
My first version of patch had `mut_self` parameter (which is the same as `exclusive`). Have a look at second commit: https://github.com/carllerche/bytes/pull/92/commits/70dbf6cdb9d3aed9af87dd861300beb5c49fa03a.

Benchmark shows nothing, because branch is cheap (in this case) even if function is not inlined.

I can revert PR to that version.

carllerche commented:
I think the boolean argument is clearer. I'd revert to that.

arthurprs commented:
I went checking and the compiler won't inline it by itself, but since it's now showing in the benchs the clearer the better.

stepancheg commented:
I've rebased a patch and returned a version with `mut_self` parameter.

Regarding benchmark, I've previously said that with batch `split_off` became 5% faster. While it is definitely faster with patch, but because variance is high, it is hard to exactly, how much faster. I believe that's because of unstable `malloc` performance.

antoyo commented:
@carllerche Is there anything new for this PR?
Thank you.

carllerche commented on src/bytes.rs:
```diff
@@ -1851,7 +1851,7 @@ impl Inner {
     /// it can be done safely. As such, this fn is not public, instead other
     /// fns will use this one while maintaining the guarantees.
-    fn shallow_clone(&self) -> Inner {
+    fn shallow_clone(&self, mut_self: bool) -> Inner {
```
Maybe add as part of the comment that `mut_self` should be set to `true` only if the caller holds a `&mut self` reference to the underlying `Bytes` value.

stepancheg commented:
* rebased patch
* updated bench to use longer buffer (just in case `Bytes` won't be upgrading short vecs to ARC on split in future version)
* updated `shallow_clone` doc mentioning `mut_self`

carllerche commented:
Thanks. I switched the shallow_clone fn to be unsafe. It probably should have been unsafe before this patch, but now it definitely should

carllerche commented:
Once CI passes, I will merge.
